symbolic-debuginfo/src/dwarf.rs

```diff
@@ -1356,6 +1359,18 @@ impl<'s> Iterator for DwarfFunctionIterator<'s> {
 
         loop {
             if let Some(func) = self.functions.next() {
+                // We have seen this iterator yield duplicate top-level function entries, which combined
+                // with recursively walking its inlinees can blow past symcache limits.
+                // We suspect the reason is that the iterator does not internally deduplicate top level
+                // functions defined in different compile units. We suspect this might be caused by link-time
+                // deduplication which merges templated code that is being generated multiple times in each
+                // compilation unit.
+                let fn_range = (func.address, func.size);
+                if self.seen_ranges.contains(&fn_range) {
+                    continue;
```

Comment:
This check is happening too late. At this time, you have already collected the function and its full inline tree into a sorted vector. You can avoid this by pushing down `&mut seen_ranges` into `DwarfUnit::functions` a few lines below.